pkg/cmd/grafana-cli/commands/mergefs.go

```diff
@@ -0,0 +1,73 @@
+package commands
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"sort"
+
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+)
+
+// MergedFS contains a slice of different filesystems that can be merged together
+type MergedFS struct {
+	filesystems []fs.FS
+}
+
+// Merge filesystems
+func Merge(filesystems ...fs.FS) fs.FS {
+	return MergedFS{filesystems: filesystems}
+}
+
+// Open opens the named file.
+func (mfs MergedFS) Open(name string) (fs.File, error) {
+	for _, filesystem := range mfs.filesystems {
+		file, err := filesystem.Open(name)
+		if err == nil {
+			return file, nil
+		}
+	}
+	return nil, os.ErrNotExist
+}
+
+// ReadDir reads from the directory, and produces a DirEntry array of different
+// directories.
+//
+// It iterates through all different filesystems that exist in the mfs MergedFS
+// filesystem slice and it identifies overlapping directories that exist in different
+// filesystems
+func (mfs MergedFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	dirs := make([]fs.DirEntry, 0)
+	for _, filesystem := range mfs.filesystems {
+		if fsys, ok := filesystem.(fs.ReadDirFS); ok {
+			dir, err := fsys.ReadDir(name)
+			if err != nil {
+				return nil, err
+			}
+			dirs = append(dirs, dir...)
```

Comment:
Say we have two filesystems, both of which implement `fs.ReadDirFS`

```
# fs1

a/
a/b/
a/b/foo.cue
```

```
# fs2

a/
a/b/
a/b/bar.cue
```

When visiting the name `a/`, will this not produce duplicate entries for `a/b`?

Reply:
Simple fix for this is to use a `map[string]bool` to store the intermediate list of deduplicate directory entries as they're encountered, then just before final return, convert into a slice and sort it.

(map value type doesn't actually matter, bool is just easy to work with)

Reply:
Hmm, not sure, tested with two different filesystems but since each filepath is unique, to me it seems that the example you posted is really:

```
#fs

a/
a/b/
a/b/bar.cue
a/b/foo.cue
```

Reply:
Have a look at this ugliest thing ever - but just an example --> https://github.com/grafana/grafana/pull/33989/commits/0cba81b69a12e11788d48e2c7a235297cd544370

to be removed ofc

Reply:
Yes, having stepped through that test, it confirms my concern - in that new, dummy test, when visiting the root dir (`name = "."`), two `fs.DirEntry` with `name = "b"` are returned. The test passes because it's not actually checking for this, and `fs.WalkDir` appears to happen to internally deduplicate the paths its visited. That is, it _happens_ to work.

Amendment to my earlier suggestion - it should be a `map[string]fs.DirEntry`, so that the full item can be provided.

Also, this tells me the semantics of the `MergedFS` type (nit: `MergeFS` is more in line with Go naming conventions) here should be explicit - if multiple fs have a file at the same path, which one's will be returned? (It should really only be "earlier wins" or "later wins," and that matters because it determines whether the map entry should be overwritten or not if a duplicate is found.)